From eb41f036af8d6bcfa62b334109581b8665fa2f65 Mon Sep 17 00:00:00 2001
From: Rohit Athavale <rohit.athavale@xilinx.com>
Date: Thu, 25 May 2017 15:44:33 -0700
Subject: [PATCH 0211/1851] v4l: xilinx-dma: Update xvip_dma_init routine to
 allow for probe defer

commit 2d9103ac7e2ddea792e537494ce2fcc87a17ada4 from
https://github.com/Xilinx/linux-xlnx.git

Update xvip_dma_init() to use dma_request_chan(),
enabling probe deferral. Also update the cleanup routine
to prevent dereferencing an ERR_PTR().

Signed-off-by: Rohit Athavale <rohit.athavale@xilinx.com>
Tested-by: Rohit Consul <rohitco@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: waiting (media)
Link: https://www.spinics.net/lists/linux-media/msg133380.html
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 7bd2600cdc9a..35a1e6b01175 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -727,10 +727,12 @@ int xvip_dma_init(struct xvip_composite_device *xdev, struct xvip_dma *dma,
 
 	/* ... and the DMA channel. */
 	snprintf(name, sizeof(name), "port%u", port);
-	dma->dma = dma_request_slave_channel(dma->xdev->dev, name);
-	if (dma->dma == NULL) {
-		dev_err(dma->xdev->dev, "no VDMA channel found\n");
-		ret = -ENODEV;
+	dma->dma = dma_request_chan(dma->xdev->dev, name);
+	if (IS_ERR(dma->dma)) {
+		ret = PTR_ERR(dma->dma);
+		if (ret != -EPROBE_DEFER)
+			dev_err(dma->xdev->dev,
+				"No Video DMA channel found");
 		goto error;
 	}
 
@@ -754,7 +756,7 @@ void xvip_dma_cleanup(struct xvip_dma *dma)
 	if (video_is_registered(&dma->video))
 		video_unregister_device(&dma->video);
 
-	if (dma->dma)
+	if (!IS_ERR(dma->dma))
 		dma_release_channel(dma->dma);
 
 	media_entity_cleanup(&dma->video.entity);
-- 
2.31.1

